### PR TITLE
Add missing fields to the returned nodes from DatabricksVectorSearch

### PR DESCRIPTION
**Pull request by Jasonli1997**
# Description

The returned nodes from the DatabricksVectorSearch currently do not include `star_char`, `end_char`, and `relationships`, the last of which is required for the usage of more advanced retrieval strategies such as Auto Merging Retriever, showcased in the demo notebook. After the changes I made, the three fields above will be added to the TextNode, if the user wishes to do so by including `node_info` column when initializing the DatabricksVectorSearch object. 

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

I showed how the new feature works in the demo notebook. 

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

**Comment by adamyargerdd** [comment]
/improve
--pr_code_suggestions.extra_instructions="
Emphasize the following:
- Does the code logic cover relevant edge cases?
- Is the code logic clear and easy to understand?
- Is the code logic efficient?
- 
"
--pr_code_suggestions.num_code_suggestions_per_chunk="4"
--pr_code_suggestions.commitable_code_suggestions=true

**Comment by qian-yu-db** [comment]
@michael-berk @B-Step62 
Can you please help look into this PR? Thank you!

File: docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb [hunk 1]
```diff
@@ -192,6 +204,99 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create a vector search index\n",
+    "# it must be placed inside a Unity Catalog-enabled schema\n",
+    "\n",
+    "# We'll use self-managed embeddings (i.e. managed by LlamaIndex) rather than a Databricks-managed index\n",
```
Comment (michael-berk) [inline docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb:243]: We need an OpenAI token somewhere, right?

File: llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py [hunk 2]
```diff
@@ -216,8 +218,19 @@ def add(
         ids = []
         for node in nodes:
             node_id = node.node_id
-            metadata = node_to_metadata_dict(node, remove_text=True, flat_metadata=True)
 
+            # unroll metadata in node_content
+            node_metadata = node_to_metadata_dict(
+                node, remove_text=True, flat_metadata=True
+            )
+            node_content = json.loads(node_metadata.get("_node_content", None))
```
Comment (michael-berk) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py:226]:
```suggestion
            node_content = json.loads(node_metadata.get("_node_content", {}))
```

File: llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py [hunk 3]
```diff
@@ -333,17 +346,30 @@ def query(self, query: VectorStoreQuery, **kwargs: Any) -> VectorStoreQueryResul
...
                 col: value
                 for col, value in zip(columns[:-1], result[:-1])
-                if col not in [self._primary_key, self.text_column]
+                if col not in [self._primary_key, self.text_column, "doc_id"]
             }
-            metadata[self._primary_key] = doc_id
+            metadata = {
+                col: value
+                for col, value in metadata_dict.items()
+                if col not in ["node_info"]
```
Comment (michael-berk) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py:359]:
```suggestion
                if col != "node_info"
```

File: docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb [hunk 4]
```diff
@@ -192,6 +204,99 @@
...
     "print(response.response)"
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Advanced Usage (Auto Merging Retriever)\n",
+    "\n",
+    "This feature is enabled by the addition of node relationships to the index.\n",
```
Comment (michael-berk) [inline docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb:214]: Can you clarify why users would use this? A simple explanation of the feature and why it's beneficial would be nice

**Comment by michael-berk** [comment]
LGTM with a few minor comments! Let's also have @B-Step62 review though.

Can you also make these unrelated improvements:
* Can you update line 16 to update that you can authenticate the VectorSearchClient outside of a Databricks runtime with workspace url and token - [docs](https://api-docs.databricks.com/python/vector-search/databricks.vector_search.html#databricks.vector_search.client.VectorSearchClient)
  * Line is `"Install llama-index and databricks-vectorsearch. You must be inside a Databricks runtime to use the Vector Search python client."`
* Can we update line 131 to leverage [create_endpoint_and_wait](https://api-docs.databricks.com/python/vector-search/databricks.vector_search.html#databricks.vector_search.client.VectorSearchClient.create_endpoint_and_wait) instead of `create_endnpoint`?

File: docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb [hunk 5]
```diff
@@ -64,6 +64,11 @@
     "    ServiceContext,\n",
     "    StorageContext,\n",
     ")\n",
+    "from llama_index.core.storage.docstore import SimpleDocumentStore\n",
+    "from llama_index.core.node_parser import HierarchicalNodeParser, get_leaf_nodes\n",
+    "from llama_index.core.retrievers import AutoMergingRetriever\n",
+    "from llama_index.core.query_engine import RetrieverQueryEngine\n",
```
Comment (B-Step62) [inline docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb:70]: Can you move these imports to the `Advanced Usage` section? These are required only for the particular advanced use case.

File: docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb (hunk 4 above)
Comment (B-Step62) [reply docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb:214]: Also this appears to be a bit misleading - this sounds like AutoMergingRetriever is a feature of the DatabricksVectorSearch which is automatically enabled wen the index contains node relationship. Can we rephrase this to clarify (1) this is feature of `AutoMergingRetriever` class (2) setting node relationship is prerequisite.

File: docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb [hunk 6]
```diff
@@ -192,6 +204,99 @@
...
+    "docstore.add_documents(nodes)\n",
+    "\n",
+    "databricks_vector_store = DatabricksVectorSearch(\n",
+    "    index=databricks_index,\n",
+    "    text_column=\"text\",\n",
+    "    columns=[\n",
+    "        \"node_info\"\n",
+    "    ],  # YOU MUST ALSO RECORD YOUR METADATA FIELD NAMES HERE\n",
+    ")  # text_column is required for self-managed embeddings\n",
+    "# NOTE: node_info is REQUIRED for the node relationships to be retrieved from the index\n",
```
Comment (B-Step62) [inline docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb:272]: nit: Can you clean up the comments format to align with standard style - put a comment above a respective parameter, like other code snippet.

File: docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb [hunk 7]
```diff
@@ -192,6 +204,99 @@
...
+    "    primary_key=\"my_primary_key_name\",\n",
+    "    embedding_dimension=1536,  # match the embeddings model dimension you're going to use\n",
+    "    embedding_vector_column=\"my_embedding_vector_column_name\",  # you name this anything you want - it'll be picked up by the LlamaIndex class\n",
+    "    schema={\n",
+    "        \"my_primary_key_name\": \"string\",\n",
+    "        \"my_embedding_vector_column_name\": \"array<double>\",\n",
+    "        \"text\": \"string\",  # one column must match the text_column in the DatabricksVectorSearch instance created below; this will hold the raw node text,\n",
+    "        \"doc_id\": \"string\",  # one column must contain the reference document ID (this will be populated by LlamaIndex automatically)\n",
+    "        # add any other metadata you may have in your nodes (Databricks Vector Search supports metadata filtering)\n",
+    "        # NOTE THAT THESE FIELDS MUST BE ADDED EXPLICITLY TO BE USED FOR METADATA FILTERING\n",
```
Comment (B-Step62) [inline docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb:256]: Does this comment apply to `node_info` field? IIUC `node_info` is not meant to be used for metadata filtering. Also let's clearly mention the field name `node_info` cannot be changed as it is hard-coded in our VS class.

File: llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py [hunk 8]
```diff
@@ -333,17 +346,30 @@ def query(self, query: VectorStoreQuery, **kwargs: Any) -> VectorStoreQueryResul
...
+            }
+            node_info = json.loads(
+                metadata_dict.get("node_info", "{}").replace("'", '"')
+            )
```
Comment (B-Step62) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py:363]:
```suggestion
            node_info = ast.literal_eval(metadata_dict.get("node_info", "{}"))
```

The `node_info` is stringfied dictionary so `literal_eval` should be more robust. Replacing single quote `'` and treat it as a json may not always work.

Alternatively, if we want to load it as a json, we should use `json.dumps` at L232.

File: llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py (hunk 2 above)
Comment (Jasonli1997) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py]: Changes made as suggested

File: llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py (hunk 3 above)
Comment (Jasonli1997) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py]: Changes made as suggested in "minor fixes"

File: docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb (hunk 4 above)
Comment (Jasonli1997) [reply docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb]: Added a bit more context and clarified how advanced features were merely due to the addition of node relationships in the index

File: docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb (hunk 6 above)
Comment (Jasonli1997) [reply docs/docs/examples/vector_stores/DatabricksVectorSearchDemo.ipynb]: Changed as suggested in "changed doc in notebook"

**Comment by RenoDakota** [comment]
UNSUBSCRIBEOn Nov 20, 2024, at 1:32 PM, Jason Li ***@***.***> wrote:﻿

LGTM with a few minor comments! Let's also have @B-Step62 review though.
Can you also make these unrelated improvements:


Can you update line 16 to update that you can authenticate the VectorSearchClient outside of a Databricks runtime with workspace url and token - docs

Line is "Install llama-index and databricks-vectorsearch. You must be inside a Databricks runtime to use the Vector Search python client."



Can we update line 131 to leverage create_endpoint_and_wait instead of create_endnpoint?



Both are addressed in the new notebook

File: llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py [hunk 9]
```diff
@@ -216,8 +219,19 @@ def add(
...
         for node in nodes:
             node_id = node.node_id
-            metadata = node_to_metadata_dict(node, remove_text=True, flat_metadata=True)
 
+            # unroll metadata in node_content
+            node_metadata = node_to_metadata_dict(
+                node, remove_text=True, flat_metadata=True
+            )
+            node_content = json.loads(node_metadata.get("_node_content", {}))
+            metadata = node_content.get("metadata", {})
```
Comment (B-Step62) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py:228]: I'm a bit confused here. What the purpose of these conversions? We don't we access the attribute of `node` directly?

```
metadata = node.metadata or {}
node_info = {key: getattr(key) for key in self._node_info_columns}
```
It look like the json serde here is unnecessary overhead.

File: llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py [hunk 10]
```diff
@@ -152,6 +154,7 @@ def __init__(
         self._delta_sync_index_spec = index_description.delta_sync_index_spec
         self._direct_access_index_spec = index_description.direct_access_index_spec
         self._doc_id_to_pk = {}
+        self._node_info_columns = ["relationships", "start_char_idx", "end_char_idx"]
```
Comment (B-Step62) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py:157]: nit: Let's define constants for these column names and "doc_id". They are used multiple times and typo prone.

File: llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py [hunk 11]
```diff
@@ -333,17 +347,26 @@ def query(self, query: VectorStoreQuery, **kwargs: Any) -> VectorStoreQueryResul
...
             score = result[-1]
             node = TextNode(
-                text=text_content, id_=doc_id, metadata=metadata
-            )  # TODO star_char, end_char, relationships? https://github.com/run-llama/llama_index/blob/main/llama-index-integrations/vector_stores/llama-index-vector-stores-pinecone/llama_index/vector_stores/pinecone/base.py
-
+                text=text_content,
+                id_=doc_id,
+                metadata=metadata,
+                start_char_idx=node_info.get("start_char_idx", None),
+                end_char_idx=node_info.get("end_char_idx", None),
```
Comment (B-Step62) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-databricks/llama_index/vector_stores/databricks/base.py:367]:
```suggestion
                start_char_idx=node_info.get("start_char_idx"),
                end_char_idx=node_info.get("end_char_idx"),
```
nit

**Comment by logan-markewich** [comment]
@B-Step62 @Jasonli1997 are we still working on this?
